Ready-to-be-reviewed version of Blowers Masel code

12Chao opened the pull request:
<!-- Thanks for contributing code! Please include a description of your change and check your PR against the list below. For further questions, refer to the contributing guide (https://github.com/Cantera/cantera/blob/master/CONTRIBUTING.md). -->

**Checklist**

- [ ] There is a clear use-case for this code change
- [ ] The commit message has a short title & references relevant issues
- [ ] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [ ] The pull request is ready for review

**If applicable, fill in the issue number this pull request is fixing**

Fixes #

**Changes proposed in this pull request**

-
-
-

speth committed:
[Input] Add a basic YAML emitter for AnyMap / AnyValue

Commit:
[Kinetics] create ReactionFactory

Commit:
[Kinetics] add Reaction::type method

ischoegl committed:
[Kinetics] remove magic numbers from Python interface

- replace 'wrapReaction' by 'Reaction.wrap' method
- eliminate hard-coded object instantiation

Commit:
[Kinetics] add setup function wrappers to ReactionFactory

Commit:
[Kinetics] simplify newReaction

ischoegl committed:
[Kinetics] make GasKinetics extendable

- deprecate magic numbers

speth committed:
[Input] Implement serialization of species thermo objects

speth committed:
[Input] Make simple ThermoPhase objects serializable

ischoegl committed:
[Kinetics] buffer available Reaction class constructors in Cython

speth committed:
[Input] Make Species objects serializable, including transport data

speth committed:
[Input] Introduce YamlWriter class for input file serialization

Currently only handles thermo, not kinetics or transport

speth committed:
[Input] Implement serialization of gas phase reactions

speth committed:
[Input] Implement serialization of interface reactions

speth committed:
[Input] Add reactions/kinetics to YamlWriter output

speth committed:
[Input] Add transport model specification to YamlWriter output

speth committed:
[Input] Improve float formatting in YAML output

speth committed:
[Input] Use stream-based YAML output method to allow formatting control

Commit:
[Kinetics] move isElectrochemicalReaction

- relocation to ReactionFactory removes function from interface

speth committed:
[Input] Enable optional use of flow mode for outputting YAML maps

speth committed:
[Input] Preserve order of items when outputting YAML maps

The default order is the order in which items are added. Items originating from
an input file come after programmatically added items, and retain their existing
ordering.

speth committed:
[Input] Retain input species thermo data

speth committed:
[Input] Add option to skip user-defined fields when writing to YAML

speth committed:
[Input] Add pathway for serialization of equation-of-state data

speth committed:
[Input] Enable serialization of Redlich-Kwong species

speth committed:
[Input] Serialize equation-of-state data for IdealSolidSolnPhase

speth committed:
[Input] Serialize equation-of-state data for LatticePhase/LatticeSolidPhase

speth committed:
[Input] Serialize equation-of-state data for StoichSubstance

speth committed:
[Input] Serialize PDSS_IonsFromNeutral

speth committed:
[Input] Serialize PDSS_ConstVol, PDSS_IdealGas, PDSS_Water

speth committed:
[Input] Serialize PDSS_HKFT

speth committed:
[Input] Implement serialization of PDSS_SSVol

speth committed:
[Input] Serialize species-specific data for DebyeHuckel model

speth committed:
[Input] YamlWriter can use separate Thermo/Kinetics/Transport objects

speth committed:
[Input] Serialize phase state using native state variables

speth committed:
[Test] Add round-trip serialization tests for thermo models

speth committed:
[Input] Fix serialization when species has no thermo model

speth committed:
[Python] Add conversion of AnyMap to native Python types

speth committed:
[Python] Add access to input data for Solution objects

speth committed:
[Python] Add access to input data for Species objects

speth committed:
Add extra information to AnyValue::getMapWhere error messages

speth committed:
[Python] Add access to input data for Reaction objects

speth committed:
Add functions for conversion from given units into a UnitSystem

speth committed:
Make AnyMap::applyUnits idempotent

ischoegl committed:
[Reaction] replace m_synonyms by addAlias

speth committed:
[Input] Implement serialization of more complex phase types

speth committed:
[Input] Provide control over ordering of YAML fields

speth committed:
[Reactor] Fix using reactor surface chemistry with YAML input

speth committed:
[Test] Use YAML input for reactor surface test

ischoegl committed:
[Kinetics] ensure correct setup routine is called for XML node

ischoegl committed:
[Kinetics] add missing header files

ischoegl committed:
[Scons] add reference to test-help after build

ischoegl committed:
[Kinetics] initial draft for custom reaction rate defined in Python

ischoegl committed:
[Kinetics] create rudimentary hooks for CustomPyReaction

ischoegl committed:
[Kinetics] expose CustomReaction to Python

ischoegl committed:
[Kinetics] implement Python constructor for CustomReaction

ischoegl committed:
[Kinetics] implement setup of CustomReaction

speth committed:
[Examples] Normalize header comments in C++ and Fortran examples

Using a consistent formatting here makes it easier to automatically
extract a title and summary for the example index on the Cantera
website.

speth committed:
[Examples] Remove outdated READMEs

speth committed:
[Examples] Fix typos in LiC6 electrode example

speth committed:
[Examples] Don't use doublereal typedef

speth committed:
[Examples] Add license info to C++ and Fortran examples

ischoegl committed:
[Kinetics] add initial unit tests

rwest commented on include/cantera/kinetics/reaction_defs.h:
```diff
@@ -76,6 +81,31 @@ const int SURFACE_RXN = 20;
+//!  formulation and using concentrations instead of activity concentrations
+//!  for its exchange current density formula.
+//!  @deprecated To be removed after Cantera 2.5.
+const int BUTLERVOLMER_NOACTIVITYCOEFFS_RXN = 25;
```
I don't think you intend to re-introduce these things that were recently removed?

rwest commented on include/cantera/kinetics/reaction_defs.h:
```diff
@@ -84,6 +114,48 @@ const int INTERFACE_RXN = 20;
+/**
+ * A global reaction. These may have non-mass action reaction orders,
+ * and are not allowed to be reversible.
+ * @deprecated To be removed after Cantera 2.5.
```
again, more stuff that was recently removed you are accidentially adding back.
Perhaps if you rebase everything onto main it'll go away.

rwest commented on include/cantera/kinetics/RxnRates.h:
```diff
@@ -429,6 +534,142 @@ class ChebyshevRate
+ *   \f[
+ *        V_P = 2w_0 (w_0 + E_0) / (w_0 - E_0)
+ *   \f]
+ * and \f$ w_0 \f$ is  theaverage of the bond dissociation energy of the 
```
```suggestion
 * and \f$ w_0 \f$ is  the average of the bond dissociation energy of the 
```

rwest commented on include/cantera/kinetics/RxnRates.h:
```diff
@@ -429,6 +534,142 @@ class ChebyshevRate
+ *        V_P = 2w_0 (w_0 + E_0) / (w_0 - E_0)
+ *   \f]
+ * and \f$ w_0 \f$ is  theaverage of the bond dissociation energy of the 
+ * bond breaking and that beingformed, which can be approximated as
```
```suggestion
 * bond breaking and that being formed, which can be approximated as
```

rwest commented on interfaces/cython/cantera/reaction.pyx:
```diff
@@ -740,6 +741,92 @@ cdef class ChebyshevReaction(Reaction):
+
+    property intrinsic_activation_energy:
+        """
+        The intrinsic activation energy *E* [J/kmol].
```
Aren't we calling it something other than E? Like Ea0 ?

rwest commented on interfaces/cython/cantera/reaction.pyx:
```diff
@@ -740,6 +741,92 @@ cdef class ChebyshevReaction(Reaction):
         return r.rate.updateRC(logT, recipT)
 
+cdef class BlowersMasel:
+    def __cinit__(self, A=0, b=0, E=0, w=0, init=True):
```
Is it just E, or is this intrinsic one called Ea0 or something?

rwest commented on interfaces/cython/cantera/test/test_kinetics.py:
```diff
@@ -1063,6 +1063,81 @@ def test_chebyshev_bad_shape_yaml(self):
+    def test_Blowers_Masel_change_enthalpy(self):
+        gas = ct.Solution('BM_example.yaml')
+        r = gas.reaction(0)
+        E0 = r.rate.intrinsic_activation_energy
```
Even though this is just a test, let's try to be consistent with names. Is it E0 or Ea0 elsewhere?

rwest commented on src/kinetics/GasKinetics.cpp:
```diff
@@ -319,6 +326,14 @@ void GasKinetics::addChebyshevReaction(ChebyshevReaction& r)
 
+void GasKinetics::addBlowersMaselReaction(BlowersMaselReaction& r)
+{
+    // m_blowersmaselindx.push_back(nReactions()-1);
```
if this shouldn't be here, remove it

rwest commented on test/data/BM_example.yaml:
```diff
@@ -0,0 +1,211 @@
+  transport: multicomponent
+  state: {T: 300.0, P: 1 atm}
+  deprecated: >-
+    The phase 'ohmech-multi' in the file 'h2o2.yaml' is deprecated and will be
```
Using deprecated stuff removed in 2.5?

rwest replied:
Can we just delete the whole `ohmech-multi` block from this yaml file?

rwest commented on test/data/BM_example.yaml:
```diff
@@ -0,0 +1,211 @@
+- equation: O + H2 <=> H + OH  # Reaction 2
+  type: Blowers-Masel
+  rate-constant: {A: 3.87e+04, b: 2.7, Ea0: 6260.0, w: 1e9}
+  note: Example would be clearer if we tweak the Ea0 so the actual Ea matches Reaction 1
```
Can we do this tweak, instead of writing in the note?  Actually, the note doesn't make sense since we don't have a "Reaction 1" currently. This as when we had two reactions, one Blowers-Masel and one normal, and we compared them.  Either add in a comparison reaction, or remove this note, and change the ` # Reaction 2` comment above.

rwest commented on test/data/surface-phases_bm.yaml:
```diff
@@ -0,0 +1,162 @@
+  sticking-coefficient: [1.0, 0.0, 0.0, 1000000]
+
+
+graphite-anode-species:
```
do we need all this graphite anode stuff in the minimal example?

rwest commented on data/BM_ptcombust.yaml:
```diff
@@ -0,0 +1,228 @@
+  ****                                                                  *
+  ****     CH4-O2 SURFACE MECHANISM  ON PT                              *
+  ****                                                                  *
+  ****     Version 1.2   November  1995                                 *
```
If this has been changed, by turning things into Blowers masel, please update the header to make that clear. Don't want someone to come across this file and think it is the Deutcschmann mechanism, which it isn't.
 Maybe remove other stuff to make a minimal example.

rwest commented on data/BM-ptcombust-Motz-Wise.yaml:
```diff
@@ -0,0 +1,230 @@
+  ****                                                                  *
+  ****     Version 1.2   November  1995                                 *
+  ****                                                                  *
+  ****     O. Deutschmann, IWR, Heidelberg University, Germany          *
```
Same comment as below - if you've changed the mechanism, please change the header. Perhaps change the mechanism a LOT to make it clearly different, eg. delete irrelevant stuff.

rwest reviewed (commented):
Thanks Chao - nice work. 
I was pressed for time so this review is not super thorough. I might have missed stuff. But I did make a few suggestions.

General themes
1. I think you reintroduce some code that was recently removed. Perhaps just rebasing onto the latest main will fix this.
2. Please try to be consistent with parameter names. Is it E or E0 or Ea0 ?
3. When including test mechanisms, make them a minimal with just the stuff needed to test it, or a *real* model, but not some hacked hybrid that looks like a real model but has arbitrary changes. Someone may come across it and think it's the real thing, without knowing about the changes.  I think in this case we want the minimal models. Once we're actually *using* BM rates we may have some real models to include as examples.
4.  Once finished with the code, probably want to interactive rebase, or reset everything and create a few new commits, so the whole thing is a few clear commits, rather than a lot of separate fixes and changes. Cantera style is to have a clean, linear, history, where each commit finishes the job it sets out to do.

ischoegl committed:
[Kinetics] introduce boolean to mark reaction validity

ischoegl committed:
[Kinetics] replace magic numbers

ischoegl committed:
[Kinetics] create hooks in GasKinetics

ischoegl committed:
[Kinetics] simplify naming of CustomPyRate

ischoegl committed:
[Kinetics] add default constructor for C++ Reaction

ischoegl committed:
[Kinetics] mark unused reaction type magic numbers

ischoegl committed:
[Kinetics] Replace remaining references to reaction_type

ischoegl committed:
[Kinetics] expose reaction_type_str to Python

ischoegl committed:
[Kinetics] deprecate functions referencing magic numbers

ischoegl committed:
[Kinetics] finalize deprecation of reaction type magic numbers

ischoegl committed:
[Kinetics] make C++ pure abstract base class

ischoegl committed:
[Func1] make CxxFunc1 shareable

ischoegl committed:
[Kinetics] make CustomRate persistent withing CustomReaction

ischoegl committed:
[Kinetics] Create C++ abstract base class RxnRate

ischoegl committed:
[Kinetics] enable calculation of rate constants for CustomReaction

ischoegl committed:
[Kinetics] generalize addition of RxnRate in GasKinetics

ischoegl committed:
[Kinetics] add alternative ArrheniusRate

ischoegl committed:
[Kinetics] update state used for RxnRate calculations

ischoegl committed:
[UnitTest] Create stand-alone file to test reactions

ischoegl committed:
[Func1] prevent invalid callback functions.

ischoegl committed:
[Kinetics] create TestReaction for benchmarking purposes

ischoegl committed:
[UnitTest] include new test in __init__

ischoegl committed:
[Kinetics] expose TestReaction to Python interface

ischoegl committed:
[UnitTest] add tests for TestReaction

ischoegl committed:
[samples] illustrate usage of Python CustomReaction

ischoegl committed:
[Kinetics] minor simplifications and edits

rwest commented on data/BM-ptcombust-Motz-Wise.yaml:
```diff
@@ -0,0 +1,115 @@
+  type: surface-Blowers-Masel
+  sticking-coefficient: {A: 0.023, b: 0, Ea0: 0, w: 1000000}
+  Motz-Wise: false
+- equation: OH + PT(S) => OH(S)  # Reaction 10
```
```suggestion
- equation: OH + PT(S) => OH(S)
```

rwest replied:
etc.

rwest commented on data/BM_ptcombust.yaml:
```diff
@@ -0,0 +1,111 @@
+- equation: O2 + 2 PT(S) => 2 O(S)  # Reaction 5
+  type: surface-Blowers-Masel
+  sticking-coefficient: {A: 0.023, b: 0, Ea0: 0, w: 1000000}
+- equation: OH + PT(S) => OH(S)  # Reaction 10
```
```suggestion
- equation: OH + PT(S) => OH(S)
```
etc.

rwest commented on include/cantera/kinetics/RxnRates.h:
```diff
@@ -84,6 +84,110 @@ class Arrhenius
+     */
+    doublereal updateRC(doublereal logT, doublereal recipT, doublereal deltaH) {
+        m_E = activationEnergy_R(deltaH);
+        return m_A * std::exp(m_b*logT - m_E*recipT);
```
```suggestion
        return m_A * std::exp(m_b * logT - m_E * recipT);
```

rwest commented on interfaces/cython/cantera/reaction.pyx:
```diff
@@ -740,6 +741,92 @@ cdef class ChebyshevReaction(Reaction):
         return r.rate.updateRC(logT, recipT)
 
+cdef class BlowersMasel:
+    def __cinit__(self, A=0, b=0, Ea0=0, w=0, init=True):
```
```suggestion
    def __cinit__(self, A=0, b=0, E0=0, w=0, init=True):
```

rwest commented on interfaces/cython/cantera/reaction.pyx:
```diff
@@ -740,6 +741,92 @@ cdef class ChebyshevReaction(Reaction):
+cdef class BlowersMasel:
+    def __cinit__(self, A=0, b=0, Ea0=0, w=0, init=True):
+        if init:
+            self.rate = new CxxBlowersMasel(A, b, Ea0 / gas_constant, w / gas_constant)
```
```suggestion
            self.rate = new CxxBlowersMasel(A, b, E0 / gas_constant, w / gas_constant)
```

rwest commented on interfaces/cython/cantera/reaction.pyx:
```diff
@@ -740,6 +741,92 @@ cdef class ChebyshevReaction(Reaction):
+
+    property intrinsic_activation_energy:
+        """
+        The intrinsic activation energy *Ea0* [J/kmol].
```
```suggestion
        The intrinsic activation energy *E0* [J/kmol].
```

rwest commented on interfaces/cython/cantera/reaction.pyx:
```diff
@@ -740,6 +741,92 @@ cdef class ChebyshevReaction(Reaction):
         r.rate.update_C(&logP)
         return r.rate.updateRC(logT, recipT)
 
+cdef class BlowersMasel:
```
Should this class have docsctrings? (do other rate classes?)

rwest commented on interfaces/cython/cantera/reaction.pyx:
```diff
@@ -740,6 +741,92 @@ cdef class ChebyshevReaction(Reaction):
+            return self.rate.activationEnergy_R0() * gas_constant
+
+    def __repr__(self):
+        return 'BlowersMasel(A={:g}, b={:g}, Ea0={:g}, w={:g})'.format(
```
```suggestion
        return 'BlowersMasel(A={:g}, b={:g}, E0={:g}, w={:g})'.format(
```

rwest commented on interfaces/cython/cantera/reaction.pyx:
```diff
@@ -810,6 +897,74 @@ cdef class InterfaceReaction(ElementaryReaction):
             r.sticking_species = stringify(species)
 
+cdef class BMInterfaceReaction(BlowersMaselReaction):
+    """ A reaction occurring on an `Interface` (i.e. a surface or an edge) """
```
add to this docstring explaining it's a Blowers Masel reaction

rwest commented on interfaces/cython/cantera/test/test_kinetics.py:
```diff
@@ -1063,6 +1063,81 @@ def test_chebyshev_bad_shape_yaml(self):
+        w = r.rate.bond_energy
+        A = r.rate.pre_exponential_factor
+        b = r.rate.temperature_exponent
+        vp = 2 * w * (w+E0) / (w - E0)
```
```suggestion
        vp = 2 * w * (w + E0) / (w - E0)
```

rwest commented on interfaces/cython/cantera/test/test_kinetics.py:
```diff
@@ -1195,6 +1298,26 @@ def test_modify_chebyshev(self):
+        R.rate = ct.BlowersMasel(A2, b2, Ta_intrinsic, w)
+        Ta2 = R.rate.activation_energy(gas.delta_enthalpy[0]) / ct.gas_constant
+        gas.modify_reaction(0, R)
+        self.assertNear(A2*T**b2*np.exp(-Ta2/T), gas.forward_rate_constants[0])
```
```suggestion
        self.assertNear(A2 * T**b2 * np.exp(-Ta2 / T), gas.forward_rate_constants[0])
```

rwest commented on src/kinetics/InterfaceKinetics.cpp:
```diff
@@ -656,6 +695,99 @@ SurfaceArrhenius InterfaceKinetics::buildSurfaceArrhenius(
+            const ThermoPhase& p = thermo(iPhase);
+            size_t k = p.speciesIndex(sp.first);
+            if (sp.first == sticking_species) {
+                multiplier *= sqrt(GasConstant/(2*Pi*p.molecularWeight(k)));
```
```suggestion
                multiplier *= sqrt(GasConstant / (2 * Pi * p.molecularWeight(k)));
```

rwest commented on test/data/BM_example.yaml:
```diff
@@ -0,0 +1,213 @@
+    geometry: atom
+    well-depth: 136.5
+    diameter: 3.33
+- name: N2
```
Delete things that aren't used, like N2, Ar, H2O2, etc. etc. to keep the file "minimal"

rwest reviewed (commented):
Few more suggestions, then good to go! Thank you

12Chao committed:
[Kinetics/Cython] Add Blowers Masel reaction class in GasKinetics

This commit inclues Blowers Masel rate class and Blowers
Masel reaction class, which are aimed to be used in GasKinetics.
The changes are implemented in both C++ and Python/Cython interfaces.

12Chao committed:
[Test/Cython] Add unit tests for Blowers Masel reaction and rate classes

The C++ test is only for Blowers Masel rate test, and the tests
in Python interface are for both the reaction and rate classes.
Two minimal yaml files are added as the input file for the tests

12Chao committed:
[Kinetics/Cython] Add Blowers Masel interface reaction and
reaction rate classes in InterfaceKinetics

The changes are made in both C++ and Cython interfaces.

12Chao committed:
[Test/Cython] The unit tests for Blowers Masel interface
reaction and reaction rates classes

The tests are added in both C++ and Python interfaces. Two
minimal yaml files are included for the tests.

speth committed:
[Input] Enable conversion of added data to units of the parent AnyMap

speth committed:
[Input] Share UnitSystems among nested AnyMaps

Since nested maps will often have the same units as their parents,
they can share the same UnitSystem instance until a change is made to
one map, at which point a new UnitSystem will be created for that map
and its children.

This reduces the memory footprint of storing UnitSystem objects, and
will become more important as the UnitSystem object starts to carry
additional data.

ischoegl committed:
[Kinetics] eliminate reg_addRxn and reg_modRxn

ischoegl committed:
[Kinetics] make RxnRate thread-safe

ischoegl committed:
[Kinetics] create State data container for RxnRate evaluations

ischoegl committed:
[examples] switch custom_reactions.py to gri30.yaml

ischoegl committed:
[Kinetics] mark derived RxnRate classes as 'final'

ischoegl committed:
[Kinetics] switch from map to vector of RxnRate pointers

ischoegl committed:
[Kinetics] store specialized reaction rate types separately

ischoegl committed:
[Kinetics] store specialized RxnRate objects without pointers

ischoegl committed:
[kinetics] add signatures for analytical derivatives to RxnRates

speth committed:
[Input] Add control over units of constructed AnyMap objects

The specified units will be used when converting quantities to scalar
values with the applyUnits() function, and will be output to a 'units'
section when serializing to YAML.

speth committed:
[Input] Add option of setting units to YamlWriter

speth committed:
[Input] Handle variable output units for all reaction types

speth committed:
Eliminate redundancy in implementation of AnyValue::as

speth committed:
[Input] Handle variable output units for all thermo/species thermo types

speth committed:
[Input] Add tests of YamlWriter for surface and edge phases

speth committed:
[Input/Python] Enable generation of YAML files from Python Solutions

ischoegl committed:
[Kinetics] address 'simple' review comments for new reaction framework

* streamline names:
  - RxnRate -> ReactionRate
  - CustomPyRate -> CustomFunc1Rate
  - CustomPyReaction -> CustomFunc1Reaction
* label some (potentially temporary) methods as @internal
* minor fixes in Python interface

ischoegl committed:
[Kinetics] introduce reaction-rate specific shared data types

speth committed:
[Input] Serialize Redlich-Kwong binary interaction parameters

ischoegl committed:
[Kinetics] create ArrheniusRate from AnyMap

ischoegl committed:
[Kinetics] add XML related 'todo' comments to ReactionFactory

After removal of XML support, some service functions can be removed to
simplify the code structure.

ischoegl committed:
[Kinetics] implement Reaction::setup method

The member 'setup' method avoids having to recreate separate stand-alone
setup methods.

ischoegl committed:
[Kinetics] add MultiRateBase evaluator class

ischoegl committed:
[Kinetics] switch to MultiRateBase framework

ischoegl committed:
[Kinetics] minor updates

ischoegl committed:
[Kinetics] add intermediate class Reaction2 as base for new framework

The intermediate class prevents naming clashes between derived classes
(e.g. 'rate' members).

ischoegl committed:
[1D] fix file output if radiation_enabled=True

speth committed:
[CI] Manually install Boost headers on Windows

GitHub has removed Boost from their default images. See
https://github.com/actions/virtual-environments/issues/2667

Fixes #990

ischoegl committed:
[Func1] roll back blocking method callbacks

ischoegl committed:
[Kinetics] implement review comments

ischoegl committed:
[Kinetics] finalize ReactionRate and corresponding data types

ischoegl committed:
[Kinetics] simplify Reaction.wrap

ischoegl committed:
[Kinetics] rename Reaction2::setup to Reaction2::setParameters

ischoegl committed:
[cti2yaml] fix conversion for transport='None'

speth committed:
Fix missing include in ReactionData.h

Fixes #999

speth committed:
[CI] Run one build with PCH disabled

12Chao committed:
reduce the test input files and modify the tests to keep them working

12Chao committed:
change 'BMInterfaceReaction' to 'BlowersMaselInterfaceReaction'

12Chao committed:
make the type of surface Blowers-Masel reaction to be specified without surface in the surface phase in yaml file

12Chao committed:
delete memeber variable 'm_E'

12Chao committed:
fix the the syntax error in doc strings to make it rendered correctly by doxygen

12Chao committed:
add documentation about Blowers Masel in gas and interface phases

leesharma committed:
Fix documentation bug for YAML format

Makes the `Reaction.listFromFile` example valid by adding the required
Kinetics object as the second argument. Now the example will run as-is and
includes a sentence stating that the `Kinetics` object is required with YAML.

Fixes #1001

12Chao committed:
Adjust the definition of the average bond energy w and the input data type of Blowers-Masel reaction rates

12Chao committed:
add more docstrings about the Blowers-Masel classes

speth committed:
Fix some formatting/whitespace issues

bryanwweber committed:
Simplify header includes in BandMatrix

If CT_USE_LAPACK is false, then CT_SUNDIALS_USE_LAPACK is also always
false. Thus the CT_SUNDIALS_USE_LAPACK branch here was never being
exercised.

bryanwweber committed:
[SCons] Fix LAPACK defines for SUNDIALS

In SUNDIALS >= 5.5, there is no longer a define that indicates SUNDIALS
was compiled with LAPACK globally. Instead, the specific SUNLINSOL
libraries that actually use LAPACK are listed explicitly.

bryanwweber committed:
[SCons] Don't print Boost info if it isn't found

bryanwweber committed:
Fix compiler warning with more recent SUNDIALS

SUNDIALS 4.0 deprecated the SUNSPGMR function name, replacing it with
a more specific version. The more specific version isn't available in
SUNDIALS 3.0 though, so we need another check here.

12Chao committed:
Add an example about use case of Blowers-Masel code

bryanwweber committed:
[SCons/Samples] Fix finding OMP library

It is possible for the OMP header to be installed in the search
path, but the library not to be available. This can happen, for
example on macOS when Boost and GCC are installed by Homebrew,
the Homebrew include path is set to find Boost, and the Apple
compilers are being used.

ischoegl committed:
Update to 88 character maximum per line

ischoegl committed:
Correct reference to CI runners

band-a-prend committed:
Fix python_minimal interface to install Cantera YAML tools

The python_minimal SConscript did't copy
cantera yaml tools to python_minimal directory
and therefore didn's install these scripts.

Signed-off-by: band-a-prend <torokhov-s-a@yandex.ru>

speth committed:
[Input] Set rate_units for reactions whenever possible

Setting the rate_units for a Reaction requires a Kinetics object. When created from
a YAML file, this is available while the Reaction is being constructed. For Reactions
created from CTI/XML or directly in code, this can be done when the Reaction is added
to a Kinetics object. Serializing using a non-default unit system is not possible if
neither of these conditions are met.

12Chao committed:
Add 2 plots in the Blowers-Masel example

speth committed:
[Input] Fix key ordering when outputting to Python

Fixes an error in the AnyValue assignment operator introduced in dab4b7066.

speth committed:
[Input] Use full AnyMap sorting algorithm in all cases

The simpler sorting algorithm used in Python did not account for the
additional ordering rules allowed by the AnyMap class.

speth committed:
[Input] Refactor to reduce Python manipulations of AnyMap

Eliminates the 'mergeAnyMap' function, and introduces a 'parameters' method
for classes to return an AnyMap which can optionally contain the user-provided
input data, rather than needing to create the AnyMap in advance and add the
user-created fields separately.

speth committed:
[Python/Input] Make function names more Pythonic

speth committed:
[Input/Examples] Update Python examples to demonstrate YAML output

speth committed:
[Input/Test] Add tests of solution properties for round-tripped phases

speth committed:
[Input] Fix AnyMap handling of 2D string / boolean arrays

speth committed:
[Input/Test] Improve test coverage of YAML serialization

speth committed:
[Examples/Python] Reformat docstring for Blowers-Masel example

Create a shorter leading paragraph which will be shown in the index
of examples on the website.

Document the dependencies.

Modify plot formatting slightly and display them before the script exits.

speth committed:
[Kinetics] Use calculateRateCoeffUnits for sticking coefficients

speth committed:
[Input] Serialize BlowersMasel reactions

speth committed:
Bump version to 2.6.0a2

12Chao committed:
use assertNear to test the exact ratio of rate constants
before and after change the coverage dependencies

12Chao committed:
Fix issue 717
